src/core/default/mappers/simple/lambda_deployer.py

```diff
@@ -221,202 +222,381 @@ def _update_simple_lambda(
-            new_resource.external_dependencies
-        )
-        create_dependencies = new_resource.external_dependencies.difference(
-            previous_resource.external_dependencies
+
+def __update_permissions_create(
+    output_task: OutputTask,
+    role_name_output: str,
+    permission_output: frozenset[dict],
+    previous_resource: simple_xlambda.simple_function_model,
+    new_resource: simple_xlambda.simple_function_model,
+) -> frozenset[Dict]:
+
+    create_permissions = new_resource.permissions.difference(
+        previous_resource.permissions
+    )
+    if not create_permissions:
+        log.debug("Simple lambda, no new permissions")
+        return permission_output
+
+    output_task.update(comment=f"Updating Policies, creating Permissions")
+    policies: List[frozendict] = []
+    for permission in create_permissions:
+        rv = frozendict(add_policy(role_name_output, permission))
+        policies.append(rv)
+
+    permission_output = frozenset(list(permission_output) + policies)
+    log.debug(f"Simple lambda, added {len(create_permissions)} permissions")
+    return permission_output
+
+
+def __update_permissions_remove(
+    output_task: OutputTask,
+    role_name_output: str,
+    permission_output: frozenset[dict],
+    previous_resource: simple_xlambda.simple_function_model,
+    new_resource: simple_xlambda.simple_function_model,
+) -> frozenset[Dict]:
+
+    remove_permissions = previous_resource.permissions.difference(
+        new_resource.permissions
+    )
+    if not remove_permissions:
+        log.debug("Simple lambda, no permissions to remove")
+        return permission_output
+
+    output_task.update(comment=f"Updating Policies, removing Permissions")
+    for permission in remove_permissions:
+        previous_permission_output_list = [
+            x for x in permission_output if x.get("hash") == permission.hash
+        ]
+
+        if len(previous_permission_output_list) > 1:
+            raise Exception
+
+        previous_permission_output = previous_permission_output_list[0]
+
+        delete_policy(role_name_output, previous_permission_output)
+
+        permission_output = frozenset(
+            [x for x in permission_output if not x.get("hash") == permission.hash]
         )
 
-        previous_dependency_output: List = list(mutable_previous_output.get("layers"))
+    log.debug(f"Simple lambda, removed {len(remove_permissions)} permissions")
+    return permission_output
+
+
+def _update_source_code(
+    output_task: OutputTask,
+    function_name: str,
+    mutable_previous_output: Dict,
+    previous_output: Dict,
+    previous_resource: simple_xlambda.simple_function_model,
+    new_resource: simple_xlambda.simple_function_model,
+):
+
+    if previous_resource.src_code_hash == new_resource.src_code_hash:
+        log.debug("Simple lambda, source code didn't change")
+        return False
+
+    sleep(3)
```

Comment:
are those sleep functions really necessary?

Reply:
Since this `mapper` is the most complex as far as doing multiple API calls, there are a lot of times that it needs wait on previous operations to finish deploying completely in the cloud. For example, when updating the IAM permission of the function, the API returns but it takes AWS time to propagate the change to all regions. 

This is definitely a hacky way of doing it, but at some point, we are going to have to rearchitect how this mapper works to be more systematic and allow recovery from partial failed deployment states.